src/ir/access_path_selectors_set.h

```diff
@@ -0,0 +1,91 @@
+// unique anyway. We lazily unique the entries when it matters.
+class AccessPathSelectorsSet {
+ public:
+  // The default constructor will construct an empty set.
+  AccessPathSelectorsSet() = default;
+
+  // Allow creating an AccessPathSelectorSet with an explicit list of items
+  // that should be in that set.
+  explicit AccessPathSelectorsSet(std::vector<AccessPathSelectors> contents)
+    : inner_vec_(std::move(contents)) {}
+
+  // Returns a set that is the union of the two passed-in sets.
+  static AccessPathSelectorsSet Union(
+      AccessPathSelectorsSet set1, AccessPathSelectorsSet set2) {
+    set1.inner_vec_.insert(
+        set1.inner_vec_.end(),
+        std::make_move_iterator(set2.inner_vec_.begin()),
+        std::make_move_iterator(set2.inner_vec_.end()));
+    return set1;
+  }
+
+  // Returns a set that is the intersection of the two passed-in sets.
+  static AccessPathSelectorsSet Intersect(
+      AccessPathSelectorsSet set1, AccessPathSelectorsSet set2) {
+    AccessPathSelectorsSet result;
+
+    // Make the first set into a hash set for efficient lookup.
+    absl::flat_hash_set<AccessPathSelectors> hash_set1 =
+      AccessPathSelectorsSet::CreateAbslSet(std::move(set1));
+
+    // Place items from the second set into the result only if they were in
+    // the first set.
+    for (AccessPathSelectors &path : set2.inner_vec_) {
+      if (hash_set1.contains(path)) {
+        result.inner_vec_.push_back(std::move(path));
+      }
+    }
+
+    return result;
+  }
+
+  // Returns a set that has the same elements as child_set but with
+  // parent_selector added as a parent to each of them.
+  static AccessPathSelectorsSet AddParentToAll(
+      Selector parent_selector, AccessPathSelectorsSet child_set) {
+    AccessPathSelectorsSet result;
+
+    for (AccessPathSelectors &path : child_set.inner_vec_) {
+      result.inner_vec_.push_back(
+          AccessPathSelectors(parent_selector, std::move(path)));
+    }
+
+    return result;
+  }
+
+  // Move the contents of the given AccessPathSelectorsSet into a new
+  // absl::flat_hash_set. This uniques and provides access to the contents of
+  // this "set".
+  static absl::flat_hash_set<AccessPathSelectors> CreateAbslSet(
+      AccessPathSelectorsSet original_set) {
+    return absl::flat_hash_set<AccessPathSelectors>(
+        std::make_move_iterator(original_set.inner_vec_.begin()),
+        std::make_move_iterator(original_set.inner_vec_.end()));
+  }
+
+ private:
+  // The inner vector implementing the set. It is not guaranteed that the
+  // contents of this vector are unique (although, due to the invariants of
+  // the type tree, we suspect they may well be). It is not guaranteed (and
+  // not expected) that the contents of this vector are sorted. We can get
+  // away with this only because we do not allow access to the contents of
+  // this set until it is requested that we turn it into a flat_hash_set.
+  std::vector<AccessPathSelectors> inner_vec_;
```

Comment:
Make this a list if you don't need random access?

Reply:
What kind of list did you have in mind? I chose `std::vector` because it's common, simple to reason about, and usually pretty performant due to its contiguous memory backing. `std::list` and `std::forward_list` are doubly and singly linked lists, respectively; they tend to be more unusual to encounter in C++ code and would probably be slower for reading due to the additional pointer dereferences.

Reply:
I was thinking `std::list` mainly prompted by the copy that I saw in `union`.  A splice would be more efficient there. Other than that it looks like we only ever iterate over all the elements forwards or backwards, which is also efficient?

I have to admit that I did not think if the reads would be slower. Do you think there will be additional pointer dereferences in vector vs list when iterating?

Reply:
I think I'm going to just go with `std::vector` for now, but open to changing this in a follow-up PR if there's a good reason to switch to a different inner collection.